tensorboard/plugins/graph/graph_util.py

```diff
@@ -0,0 +1,145 @@
+# Copyright 2019 The TensorFlow Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Utilities for graph plugin."""
+from tensorboard.compat.proto import function_pb2
+
+
+class _ProtoListNonUniqueKeyError(Exception):
+  pass
+
+
+class _SameKeyDiffContentError(Exception):
+  pass
+
+
+def _safe_copy_proto_list_values(dst_proto_list, src_proto_list, get_key):
+  """Safely copies the value from src_proto_list to dst_proto_list.
+
+  First checks whether keys in both dst_proto_list and src_proto_list are
+  unique. If not, it raises a _ProtoListNonUniqueKeyError. It copies if
+  dst_proto_list does not contain an item with the same key. In case an item
+  is found with the same key, it checks whether contents match and, if not, it
+  raises a _SameKeyDiffContentError.
+
+  Args:
+    dst_proto_list: A `RepeatedCompositeContainer` or
+      `RepeatedScalarContainer` into which values should be copied.
+    src_proto_list: A container holding the same kind of values as in
+      `dst_proto_list` from which values should be copied.
+    get_key: A function that takes an element of `dst_proto_list` or
+      `src_proto_list` and returns a key, such that if two elements have
+      the same key then it is required that they be deep-equal. For
+      instance, if `dst_proto_list` is a list of nodes, then `get_key`
+      might be `lambda node: node.name` to indicate that if two nodes
+      have the same name then they must be the same node. All keys must
+      be hashable.
+
+  Raises:
+    _ProtoListNonUniqueKeyError: a proto_list contains items with duplicate
+      keys.
+    _SameKeyDiffContentError: an item with the same key has different contents.
+  """
+
+  def _assert_proto_container_unique_keys(proto_list, get_key):
+    """Asserts whether proto_list only contains unique keys.
```

Comment:
nit: You assert *that* it contains only unique keys (a judgment), not
*whether* it does (a boolean).